Allow home view to reorder

drewvolz opened the pull request:
* Adds a view to reorder list of main views on home view
* Saves order to async storage to use back on the home view
* Reorders home view with any changes made
* Adds @hawkrives' WIP async storage higher-order component

- [x] Works on Android
- [x] Works on iOS

<img width="385" alt="opening-in" src="https://cloud.githubusercontent.com/assets/5240843/20651302/c5e82e3a-b4b0-11e6-8e01-5f0e25d6c3c3.gif">

elijahverdoorn commented on views/home.js:
```diff
@@ -37,51 +44,94 @@ const views: ViewType[] = [
+const Touchable = ({children, onPress}: {onPress: () => any, children?: Element<any>}) => {
+  return Platform.OS === 'ios'
+    ? <TouchableOpacity onPress={onPress} activeOpacity={0.65}>{children}</TouchableOpacity>
+    : <TouchableNativeFeedback onPress={onPress}>{children}</TouchableNativeFeedback>
```
This is nice. Maybe we should extract the logic of `TouchableOpacity` vs `TouchableNativeFeedback` out to a component?

elijahverdoorn replied:
By this I mean in the `/components` folder so that we can use it elsewhere where we now just use `TouchableOpacity`.

drewvolz replied:
@hawkrives thoughts? Any drawbacks from extracting into our own component to represent the two?

hawkrives replied:
That's a good idea. I've got a similar component built in the BuildingHours PR. 

I'll take a look at extracting it out.

elijahverdoorn commented on views/editHome.js:
```diff
@@ -0,0 +1,207 @@
+        contentContainerStyle={styles.contentContainer}
+        data={objViews}
+        order={this.state.order}
+        onChangeOrder={this.onOrderChange}
```
I wonder if it would be better to store the order of the items in memory, then call the AsyncStorage stuff when the user hits the done button? Seems unnecessary to be writing to storage every time they change one when they only see the result when they hit done. (I might be missing a very good reason for it to work this way!)

drewvolz replied:
That is a fair suggestion, and should be easily doable. You are making a decent suggestion -- not missing any point. Thanks.

drewvolz replied:
I have tried this now -- saving it to memory, and then calling AsyncStorage stuff when the user hits the done button. In short, it is not quick enough. 

There's no downside to saving on reorder. It's fast and asynchronous, it's just saving a list of strings. There are upsides -- it saves the order on crash of that view... which is nice if you spent a bit of time thinking and reordering, or if the app suddenly dies (as is the case of Android).

You have pointed out something that needs to change as a whole, as using `componentWillReceiveProps()` may not be the correct way to do this... it may not fire every time. 

**So** the correct answer here is "use redux or some other method of lifting application state above react", as confirmed by @hawkrives

elijahverdoorn reviewed (commented):
Looks really good overall, just a few suggestions/questions in the review.

hawkrives committed:
add redux packages

drewvolz committed:
Fixing lint errors

drewvolz committed:
Allow home view to reorder

* Add view to reorder list of main views on home view
* Save order to asynch storage to use back on the home view
* Reorder home view with any changes made
* Adds @hawkrive's WIP asynch storage higher-order component

integrate non-global-variable onDismiss function stuff

fix savedOrder check

clarify type of 'route' prop

split up components in home.js

rename 'views' to 'allViews'

extract verbose icons into smaller components

tweaks

fix prop for MenuIcon

remove borderRadius prop

clean up some types

fix android homescreen buttons

fix eslint error

clean up some styles in the stylesheet

drewvolz committed:
Remove flex to fix scrolling

drewvolz committed:
Remove whitespace and commented out code

drewvolz committed:
Patch to remote sortable-list-view

hawkrives committed:
integrate redux

hawkrives committed:
extract views list into its own file

hawkrives committed:
fix lint errors

hawkrives committed:
fix type annotation in onChangeOrder callback from EditHomeView

drewvolz committed:
Easing animations for opacity and elevation

drewvolz committed:
Bump patched rn-sortable-list hash

hawkrives reviewed (approved):
We're positive that the stock Android crashes are from a React Native bug.

This workaround's not _perfect_, but it's workable, IMHO.
